apps/req.c

```diff
@@ -1734,10 +1734,6 @@ static int do_sign_init(EVP_MD_CTX *ctx, EVP_PKEY *pkey,
 
     ret = 1;
  err:
-#ifndef OPENSSL_NO_SM2
-    if (!ret)
```

Comment:
Where is the other free?
If it happens because of EVP_MD_CTX_set_pkey_ctx(ctx, pctx); then you may still need this call fro some paths..

Reply:
Any SM2 related `free` should all go in the outer `do_X509_sign`, this should only apply to SM2 keys

Reply:
Ahh, I get you point. So it seems the proper way is to handle to outer free instead of this one...

Reply:
There are multiple issues with this currently.
I dont think you should remove the inner cleanup - If it allocates the data then it should clean up the data if it fails..
The outer one shouldnt just fall thru when there is an error.

Reply:
So at the outer cleanup, it should distinguish if the inner call is failed or not, only non-failed inner call should be cleanup at outer space..

Reply:
In the inner it should probably be doing...

.....
  EVP_MD_CTX_set_pkey_ctx(ctx, pctx);
  pctx = NULL;

......


if (pctx == NULL)
    EVP_MD_CTX_set_pkey_ctx(ctx, NULL);
else
   EVP_PKEY_CTX_free(pctx);

Reply:
It seems this is not necessary. Even if an error occurred after `EVP_MD_CTX_set_pkey_ctx` is called, the original logic can still free the `pctx` and EVP_PKEY_CTX_free accepts NULL input

Reply:
The point is if a function allocs memory and then assigns it to something in the MD_CTX, then MD_CTX probably shouldnt be pointing to a deallocated memory location when the pkey ctx is freed. If you use the other function then it frees the pkey and clears the pointer and flags that it uses.